tonconnect-server/TonConnect2.md

```diff
@@ -0,0 +1,533 @@
+
+TON apps may control various assets (coins, tokens, NFTs etc.) according to the publicly auditable and immutable logic performed by the blockchain. This way users do not have to place trust in closed private systems to not censor or steal their money.
+
+Access to TON apps is controlled by the wallets: regular applications that keep users’ cryptographic keys on their devices. Wallets enable users to sign into applications and authorize blockchain transactions.
+
+The goal of Ton Connect is to enable smooth and safe interaction between wallets, services and apps.
+
+
+### Apps vs services
+
+In this specification we use the following terminology:
+
+**App** is a software that runs on behalf of the user and does not rely on a dedicated backend server for accessing users’ data.
+
+**Service** is a software that runs on behalf of its operator, stores users’ data and regulates access to the data.
+
+For TON Connect the distinction lies in the fact that service has its own offchain identity and may request [identification of the user](#user-identification). Apps run on behalf of the user, so identifying the user does not make much sense, while app’s own identity is defined by its smart contract address.
+
+
+### Decentralized vs centralized apps
+
+In the above definitions apps are considered decentralized and services considered centralized. 
+In reality, there is a spectrum of trust-minimization and each specific app or service may rely on some centralized infrastructure.
+
+Both apps and services are often non-custodial for the security and legal reasons. Instead of managing cryptographic keys that hold users’ funds they rely on wallet apps to approve and sign transactions on behalf of the users. Wallet apps themselves are often non-custodial too: they store secret keys securely on their users’ devices and never communicate them to other computers on the network.
+
+
+### User identification
+
+To minimize cross-service tracking, TON Connect uses distinct cryptographic identities for each service.
+
+User identity is tied to the [app keypair](#app-keypair).
+
+
+### User authentication
+
+Decentralized apps authenticate users within smart contracts using message sender addresses. 
+The UI (“offchain logic”) runs on behalf of the user, so there is no need for a dedicated “sign in” scheme.
+
+Centralized apps run on the servers and may store user data. They may use “sign in” functionality to authenticate the user, where instead of username/password pair we use cryptographic keypair derived from the wallet secret seed.
+
+
+### App authentication
+
+Apps are authenticated using [TON DNS](https://ton.org/docs/#/web3/dns) protocol.
+
+Decentralized apps may use `sha256("wallet")` key for the smart contract address. 
+When the wallet receives request to sign a message to a smart contract with a given **.ton** name, a valid and up-to-date DNS record would be used to authenticate that contract address with that name.
+
```

Comment:
Is there any way in the standard for an app to use the raw contract address without a DNS record?

I can see several reasons this would make sense:

1. Owning a DNS record means that the average developer will have some wallet that keeps the NFT of the domain ownership. Wouldn't this encourage them to create a centralized failure point? This admin wallet will be able to replace the contract address at any point and steal money from users. The amount of education we will need to convince the average developer to move this domain NFT to the zero address and thus revoke control over it, will be insane. I think we need a solution that by default is trustless and does not encourage centralization.

2. There are cases where factories deploy smart contracts per user. For example in Balancer V1 on Ethereum if I remember correctly, users had to deploy a proxy contract and then access the protocol through it. The same happens with a project called https://fodl.finance. To reach isolation between borrowing accounts, this project deploys a smart contract per user. When the app deploys a new contract per user, how should the DNS angle be handled?

Reply:
I think at present stage the connect flow should work even in unauthenticated situation (without TON DNS), and when we figure out how to make DNS usable in all deployment scenarios we can slowly phase out non-authenticated uses similar to how web-browsers steer towards https.

Re: 1. In any app deployment there is a PKI burden. When you build a server you need a TLS certificate and safeguard the keys. If you do PGP you need to make your pubkey known and counter-signed by known parties. TON DNS is a convenient feature for the users and a formalized burden for developer. Developer may choose to not use TON DNS, but then they are on their own to distribute the identity of their app or contract. 

TON DNS is kinda cool because it's built on the native on-chain tech and we can imagine a decentralized version of the owner for it. E.g. your DAO/dapp contract could be made an owner of a DNS record and renewals could be done through a separate message that anyone can sent thus ensuring that the name remains attached as long as anyone among the users cares.

2: That's an interesting use-case, thanks for pointing it out! Is it possible to know whether a contract was created by another contract and could not be deployed externally (if that matters)? If so, we could support parent-child relationship similar to hierarchy of CAs in Web PKI: the parent contract has a .ton name, while children are authenticated through it by association.

Reply:
Not sure there's a PKI burden on app deployment. Our common practice is to use a temporary new PK for the deployment and dispose of it after deployment. The burden is only when this PK holds power beyond the deployment and should be managed onwards. Our deployment PK's are useless after the deploy so there's no point in keeping them

Reply:
Sorry for not being clear. I mean, you have to deal with some PKs anyway, so TON DNS does not bring anything new here. If you want it to be very decentralized you need to attach your dapp's name to some contract that no one can control, but anyone can pay up for DNS through it to keep it alive.

Reply:
As for sub-contracts I suggest adding sub-level DNS records for them (or have them _be_ such records). E.g. if you have a `mydex.ton` with separate entry points for USDTON and EURTON pairs, you may have `usdton.mydex.ton` for the first one and `eurton.mydex.ton` for the second one.

Reply:
I'm sorry, but this part of the proposal is broken. Requiring TON DNS will break the standard and make it incompatible with many dapp use-cases. TON DNS is relevant primarily for the *service* use-case. It's not appropriate at all for the *app* use-case. If you enforce the TON DNS requirement, you will prevent many dapps from being able to use TonConnect.

1. In the dapp use-case, there should be no PK to deal with. There is no entity to hold any PK. You have a few smart contracts and an open source client that anyone can run. Adding a PK where non should exist adds unnecessary complexity. Ethereum for example doesn't require dapps to deal with any PK and the dapp-wallet flow there is fine.

2. I'm trying to write a blog post to onboard new developers to TON (this is [part 1](https://society.ton.org/ton-hello-world-step-by-step-guide-for-writing-your-first-smart-contract-in-func), [part 2](https://society.ton.org/ton-hello-world-guide-for-writing-first-smart-contract-in-func-part-2), part 3 should be about a simple web client for the dapp). I can't imagine sending learning developers to register at TON DNS, an action that is a hassle, takes a week for the auction and costs money, just so they can implement the simplest client to interact with their tutorial smart contract.

3. Let's say I'm implementing a DEX and I decided to have a separate smart contract instance (and separate address) per liquidity pair. There are 50,000 such contracts for pairs deployed on Uniswap for example. Let's say I don't want to go through a router contract and I want my users to interact with these 50,000 smart contract instances directly. Why would you force me the hassle of setting up 50,000 TON DNS subdomains? It makes no sense and it adds no security.

4. Let's say the smart contract I'm implementing follows the sharded contract model like the Jetton contract. Meaning every user has a separate contract instance deployed (like on Jetton - the contract jetton-wallet). I want the dapp client to be able to send messages to these child smart contracts directly. Where am I supposed to register all the TON DNS names for these children?

5. The common case in my eyes will be like on Binance Smart Chain or Polygon, many shitty dapps that developers deploy and most of them are not very trustworthy. If you require all of them to register a TON DNS name, I guarantee that 99% of them will end up holding the NFT of the domain on their own centralized wallet. They're not going to give it up (transfer to the zero address) and you're setting them up for failure. Because by controlling the DNS record, they could replace the contract address without their users' consent and effectively steal all the money. You need to understand that it's our role as designers to create a standard of operation where making users vulnerable is not the default but the exception. We don't need to make it easy for developers to make mistakes and create super dangerous apps by accident.